### ISSUE TITLE
Rogue "wrap return expression in Ok" errors

### ISSUE THREAD

**Issue by fahrradflucht** [rust-analyzer/rust-analyzer#4584]
I can't really pinpoint the issue, so sorry for the bad description upfront 🙏.

When I open https://github.com/swc-project/swc/blob/5f261fcd5ca7c07db15632d0e62336fe0c6cf6f8/ecmascript/codegen/src/lib.rs, rust-analyzer reports a bazillion (39 to be exact 😉) errors with the message `wrap return expression in Ok`. The project compiles just fine though.

First I thought that it doesn't understand the `emit!` macros, but they expand just fine. Why is `wrap return expression in Ok` even an error? This sounds more like a code action or a fix for another error. I'm very confused...

**Comment by bjorn3** [comment]
The problem here is likely that `#[emitter]` is not expanded. You can use the following vscode configuration to enable proc-macros:

```json
{
    "rust-analyzer.cargo.loadOutDirsFromCheck": true,
    "rust-analyzer.procMacro.enable": true,
}
```

**Comment by lnicola** [comment]
I still see the errors even with proc macros enabled. The issue seems to be that the `match` statements are inferred as `()`.

**Comment by flodiebold** [comment]
> Why is wrap return expression in Ok even an error? This sounds more like a code action or a fix for another error.

Well, the error is a type mismatch, we just don't report those yet because they still have too many false positives. In the particular case where we detect a type mismatch that could be fixed by `Ok`-wrapping, we show the diagnostic and quick-fix. That usually doesn't have too many false positives, but this is one.

The `emit!` calls getting inferred as `()` seems correct from the definition. I assume that the `#[emitter]` macro does something to make these functions typecheck. We don't expand attribute-like macros yet though.

**Comment by Lonami** [comment]
I just came across this, with the following code on Visual Studio Code and the `rust-analyzer` extension:

```rust
pub fn foo() -> Result<(), ()> {
    loop {
        match bar() {
            Ok(_) => continue,
            Err(e) if true => break Ok(()),
            e => break e,
        }
    }
}

fn bar() -> Result<(), ()> {
    Ok(())
}
```

It will highlight the entire `loop`:
![image](https://user-images.githubusercontent.com/6297805/82761224-62e7d700-9df9-11ea-9d5c-8310c0bab306.png)

**Comment by flodiebold** [comment]
@Lonami that should be fixed with #4505, which should be in the next weekly release.

**Comment by ruabmbua** [comment]
Can anyone confirm, that this is already fixed? #4505 landed already.

**Comment by Lonami** [comment]
@ruabmbua the issue I highlighted in https://github.com/rust-analyzer/rust-analyzer/issues/4584#issuecomment-633268257 seems to be solved, but I don't know if OP's problem is any different or is actually the same as mine.

**Comment by edwin0cheng** [comment]
`#[emitter] is an attribute proc macro.

Closing this as a duplicate of #6029 since it's caused by an unexpanded proc macro attribute.
